scene/gui/text_edit.cpp

```diff
@@ -1019,17 +1019,34 @@ void TextEdit::_notification(int p_what) {
 					const String &str = wrap_rows[line_wrap_index];
 					int char_margin = xmargin_beg - cursor.x_ofs;
 
-					int ofs_readonly = 0;
 					int ofs_x = 0;
+					int ofs_y = 0;
 					if (readonly) {
-						ofs_readonly = cache.style_readonly->get_offset().y / 2;
 						ofs_x = cache.style_readonly->get_offset().x / 2;
+						ofs_x -= cache.style_normal->get_offset().x / 2;
+						ofs_y = cache.style_readonly->get_offset().y / 2;
+					} else {
+						ofs_y = cache.style_normal->get_offset().y / 2;
 					}
 
-					int ofs_y = (i * row_height + cache.line_spacing / 2) + ofs_readonly;
+					ofs_y += i * row_height + cache.line_spacing / 2;
 					ofs_y -= cursor.wrap_ofs * row_height;
 					ofs_y -= get_v_scroll_offset() * row_height;
 
+					if (line_wrap_index != 0) {
+						int max_size_y = get_size().height;
+						if (readonly) {
+							max_size_y -= cache.style_readonly->get_minimum_size().y;
+						} else {
+							max_size_y -= cache.style_normal->get_minimum_size().y;
+						}
+
+						if (ofs_y - row_height > max_size_y) {
+							// Line is outside the content margin, clip.
```

Comment:
Would we need to set `i = draw_amount;` to also break from the outer loop?

Reply:
Good catch, thanks! I've added that so it can break earlier when there's nothing else to draw.
Actually, I've also removed the condition `if (line_wrap_index != 0)` so it can apply the clipping on lines after line breaks too.